src/network/websockets/index.ts

```diff
@@ -10,58 +10,69 @@ type Params = {
-    socket.onclose = () => onConnectivityChange("offline");
-    socket.onerror = () =>
-      onError(
-        new Error(
-          "WebSocket error. We can’t retrieve details about the error because the browser does not provide them for security reasons."
-        )
-      );
+  const socket = new ReconnectingWebSocket(url);
+  const send = (input: {}) => {
+    socket.send(JSON.stringify(input));
+  };
+
+  return {
+    adapter: function({
+      onChange,
+      onConnectivityChange,
+      onError,
+      onPushResult
+    }) {
+      socket.onopen = () => onConnectivityChange("online");
+      socket.onclose = () => onConnectivityChange("offline");
+      socket.onerror = () =>
+        onError(
+          new Error(
+            "WebSocket error. We can’t retrieve details about the error because the browser does not provide them for security reasons."
+          )
+        );
 
-    const handlers: { [action: string]: (msg: any) => void } = {
-      ping: () => {},
-      pushResult: (msg: any) => onPushResult(msg.pushId, msg.result),
-      update: (msg: any) =>
-        onChange({
-          kind: msg.kind,
-          id: msg.id,
-          revision: msg.revision,
-          value: serialization.decode(msg.value)
-        })
-    };
-    socket.onmessage = function(event) {
-      const msg = JSON.parse(event.data);
-      const action = msg.action;
-      if (!action) return console.warn(`Unrecognized message`, msg);
-      const handler = handlers[action];
-      if (!handler) return console.warn(`Unrecognized message`, msg);
-      handler(msg);
-    };
+      const handlers: { [action: string]: (msg: any) => void } = {
+        ping: () => {},
+        pushResult: (msg: any) => onPushResult(msg.pushId, msg.result),
+        update: (msg: any) =>
+          onChange({
+            kind: msg.kind,
+            id: msg.id,
+            revision: msg.revision,
+            value: serialization.decode(msg.value)
+          })
+      };
+      socket.onmessage = function(event) {
+        const msg = JSON.parse(event.data);
+        const action = msg.action;
+        if (!action) return console.warn(`Unrecognized message`, msg);
+        const handler = handlers[action];
+        if (!handler) return console.warn(`Unrecognized message`, msg);
+        handler(msg);
+      };
 
-    const send = (input: {}) => {
-      socket.send(JSON.stringify(input));
-    };
-    return {
-      getAndObserve: (kind, id) => {
-        send({ action: "subscribe", kind, id });
-        return () => {};
-      },
-      push: async ({ kind, id, pushId, lastSeenRevision, value }) => {
-        send({
-          action: "push",
-          kind,
-          id,
-          pushId,
-          lastSeenRevision,
-          value: serialization.encode(value)
-        });
-      }
-    };
+      return {
+        getAndObserve: (kind, id) => {
+          send({ action: "subscribe", kind, id });
+          return () => {};
```

Comment:
No suggestion here, just a question, since I've never looked at this codebase before and have very little understanding of what's going on. Why is `getAndObserve()` returning a noop function? What is the idea behind `getAndObserve`, it's intended use?

Reply:
Cancelling observation is designed, but not implemented yet. Once it’s added to Oncilla, this should handle cancellation by, for example, sending an unsubscribe message:

```suggestion
          return () => send({ action: "unsubscribe", kind, id });
```

This is not related to the PR, though, just an artifact of early designs.